moveos/moveos/src/moveos.rs

```diff
@@ -231,11 +235,174 @@ impl MoveOS {
+                                .execute_function_bypass_visibility(gas_validate_func_call)?;
+
+                            return if !return_value.is_empty() {
+                                let first_return_value = return_value.get(0).unwrap();
+                                Ok(Some(
+                                    bcs::from_bytes::<bool>(first_return_value.value.as_slice())
+                                        .expect(
+                                        "the return value of gas validate function should be bool",
+                                    ),
+                                ))
+                            } else {
+                                Ok(None)
+                            };
+                        }
+
+                        Ok(None)
+                    }
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn execute_gas_charge_post(
+        &self,
+        session: &mut MoveOSSession<'_, '_, MoveOSResolverProxy<MoveOSStore>, MoveOSGasMeter>,
+        action: &VerifiedMoveAction,
+    ) -> VMResult<Option<bool>> {
+        match action {
+            VerifiedMoveAction::Function { call } => {
+                let module_id = &call.function_id.module_id;
+                let loaded_module_bytes = session.get_data_store().load_module(module_id);
+
+                let compiled_module_opt = {
+                    match loaded_module_bytes {
+                        Ok(module_bytes) => {
+                            CompiledModule::deserialize(module_bytes.as_slice()).ok()
+                        }
+                        Err(_) => {
+                            unreachable!("Module data {:} should exists", module_id.to_string());
+                        }
+                    }
+                };
+
+                let gas_free_function_info = {
+                    match compiled_module_opt {
+                        None => {
+                            return Err(PartialVMError::new(
+                                StatusCode::FAILED_TO_DESERIALIZE_RESOURCE,
+                            )
+                            .with_message(format!(
+                                "failed to deserialize module {:?}",
+                                module_id.to_string()
+                            ))
+                            .finish(Location::Module(module_id.clone())))
+                        }
+                        Some(compiled_module) => {
+                            let metadata = get_metadata_from_compiled_module(&compiled_module);
+                            match metadata {
+                                None => None,
+                                Some(runtime_metadata) => {
+                                    Some(runtime_metadata.gas_free_function_map)
+                                }
+                            }
+                        }
+                    }
+                };
```

Comment:
We can provide some functions to reuse the logic of getting things from metadata.